### ISSUE TITLE
Mandatory auth_url and 4.0

### ISSUE THREAD

**Issue by jeroenvervaeke** [ramosbugs/oauth2-rs#135]
Currently the `auth_url` is mandatory when creating a new Client: [doc](https://docs.rs/oauth2/3.0.0/oauth2/struct.Client.html#method.new)
However, auth url isn't mandatory when using the [Client Credentials Grant](https://tools.ietf.org/html/rfc6749#section-4.4) and the [Resource Owner Password Credentials Grant](https://tools.ietf.org/html/rfc6749#section-4.3).

Since you're introducing 4.0 semantic versioning would allow us to introduce a breaking change.
I propose to make the `auth_url` optional in the new method of a client, this way you don't need to enter a valid garbage url when you're using client credentials.
Before:
```rust
let client =
    BasicClient::new(
        ClientId::new("client_id".to_string()),
        Some(ClientSecret::new("client_secret".to_string())),
        AuthUrl::new("http://garbage_url".to_string())?,
        Some(TokenUrl::new("http://token".to_string())?),
    );
```
After:
```rust
let client =
    BasicClient::new(
        ClientId::new("client_id".to_string()),
        Some(ClientSecret::new("client_secret".to_string())),
        None,
        Some(TokenUrl::new("http://token".to_string())?),
    );
```

In a later version we could add a `BasicClientBuilder` struct to avoid having to provide a lot of `None` values to the `Client`

**Comment by ramosbugs** [comment]
hey @jeroenvervaeke,

this brings up an interesting API design issue that I've struggled with a few times in the past. you're correct that the authorization endpoint isn't required for the grants you mentioned. this came up with some of the newer endpoints (see #127), and we introduced a `ConfigurationError` type for those.

my worry with making the auth endpoint optional is that it'll create a situation in which code that omits it but then tries to make a request that requires it won't fail until runtime. ideally, the type system would be able to enforce this. this is already an issue for the newer endpoints like token introspection, but now it would be affecting the most common OAuth2 flows as well.

I think we'd be better off exploring an alternative solution, such as allowing a `ClientCredentialsTokenRequest` to be constructed without first constructing a `Client`.

separately, this issue just missed the cutoff for breaking changes in 4.x since I promoted the release to beta a couple of weeks ago (see #99).

thoughts?

**Comment by jeroenvervaeke** [comment]
I have a specific solution in mind that I would like to approach. I might make a POC and come back to it later.
The solution that I have in mind uses the type system to avoid incorrect configuration & avoids using the Client

**Comment by vidhatha** [comment]
@ramosbugs @ramosbugs, any updates on this? Even I am in the same situation where the `auth_url` is not required.

**Comment by ramosbugs** [comment]
This will be fixed in 5.0 (see 1d1f4d17ecdf2a3feb565eb1789cc8649cac7705 for current progress)

**Comment by ramosbugs** [comment]
This is now released in [5.0.0-alpha.1](https://crates.io/crates/oauth2/5.0.0-alpha.1). If you can, please try it out and let me know if you find any issues that should be fixed before stabilizing the 5.0 API. There's an [Upgrade Guide](https://github.com/ramosbugs/oauth2-rs/blob/main/UPGRADE.md) to help.
